File: node/src/api/backend.ts
/**
 * BackendClient — fetch-based HTTP client for the Scouter backend REST API.
 * Mirrors sdk/python/scouter/api/backend.py (all 19 methods).
 * Falls back gracefully — returns null on connection errors.
 */

import type { RegisterIntentParams, ActionPayload, MintCredentialParams } from "../models.js";

// ── Response types ───────────────────────────────────────────────────

export interface IntentResponse {
  intent_id: string;
  agent_id: string;
  natural_language: string;
  permitted_actions: string[];
  excluded_actions: string[];
  has_embedding?: boolean;
  [key: string]: unknown;
}

export interface EvaluationResponse {
  artifact_id: string;
  timestamp: string;
  intent_id: string;
  action: Record<string, unknown>;
  evaluation: {
    irreversibility_score: number;
    alignment_score: number;
    calculated_decision: string;
    actual_execution: string;
    rationale: string;
  };
  signature?: string;
  public_key_id?: string;
  [key: string]: unknown;
}

export interface SpanResponse {
  span_id: string;
  trace_id: string;
  span_type: string;
  [key: string]: unknown;
}

export interface AnalyzeTraceResponse {
  trace_id: string;
  findings: Array<{
    failure_type: string;
    severity: string;
    description: string;
    [key: string]: unknown;
  }>;
  count: number;
}

export interface ArtifactVerifyResponse {
  valid: boolean;
  artifact_id?: string;
  [key: string]: unknown;
}

export interface ComplianceExportResponse {
  export_type: string;
  total_artifacts: number;
  public_key: string;
  [key: string]: unknown;
}

export interface TelemetryStatsResponse {
  total_decisions: number;
  unique_agents: number;
  decision_breakdown: Record<string, number>;
  [key: string]: unknown;
}

export interface TelemetryRecord {
  telemetry_id: string;
  agent_id: string;
  action_type: string;
  calculated_decision: string;
  [key: string]: unknown;
}

export interface AgentStatsResponse {
  total_decisions: number;
  decision_breakdown: Record<string, number>;
  [key: string]: unknown;
}

export interface PromptAnalysisResponse {
  decision: string;
  risk_score: number;
  [key: string]: unknown;
}

export interface PromptBatchResponse {
  results: PromptAnalysisResponse[];
  [key: string]: unknown;
}

export interface CredentialResponse {
  credential_id: string;
  token?: string;
  expires_at: string;
  [key: string]: unknown;
}

export interface PbacEvalResponse {
  allowed: boolean;
  policy_id?: string;
  reason?: string;
  [key: string]: unknown;
}

export interface PbacPolicyResponse {
  policy_id: string;
  intent_id: string;
  [key: string]: unknown;
}

export interface DIDResponse {
  did: string;
  status: string;
  did_document?: Record<string, unknown>;
  private_key?: string;
  [key: string]: unknown;
}

export interface DIDKeyRotateResponse {
  new_key_id: string;
  private_key: string;
  [key: string]: unknown;
}

// ── Client ───────────────────────────────────────────────────────────

export interface BackendClientOptions {
  timeout?: number;
  apiKey?: string;
}

export class BackendClient {
  private readonly baseUrl: string;
  private readonly timeout: number;
  private readonly headers: Record<string, string>;

  constructor(baseUrl: string, options: BackendClientOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeout = options.timeout ?? 5000;
    this.headers = { "Content-Type": "application/json" };
    if (options.apiKey) {
      this.headers["X-Scouter-API-Key"] = options.apiKey;
    }
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    timeoutOverride?: number,
  ): Promise<T | null> {
    try {
      const controller = new AbortController();
      const timer = setTimeout(
        () => controller.abort(),
        timeoutOverride ?? this.timeout,
      );
      const opts: RequestInit = {
        method,
        headers: this.headers,
        signal: controller.signal,
      };
      if (body !== undefined) {
        opts.body = JSON.stringify(body);
      }
      const res = await fetch(`${this.baseUrl}${path}`, opts);
      clearTimeout(timer);
      if (!res.ok) return null;
      return (await res.json()) as T;
    } catch {
      return null;
    }
  }

  // ── Intent Registry ────────────────────────────────────────────

  async registerIntent(params: RegisterIntentParams): Promise<IntentResponse | null> {
    return this.request<IntentResponse>("POST", "/api/v1/intents", {
      agent_id: params.agentId,
      natural_language: params.naturalLanguage,
      permitted_actions: params.permittedActions,
      excluded_actions: params.excludedActions,
      principal_chain: params.principalChain ?? [],
      version: params.version ?? "1.0",
    });
  }

  async getIntent(intentId: string): Promise<IntentResponse | null> {
    return this.request<IntentResponse>("GET", `/api/v1/intents/${intentId}`);
  }

  // ── Consequence Engine ─────────────────────────────────────────

  async evaluate(
    action: ActionPayload | Record<string, unknown>,
    intentId?: string,
    traceId?: string,
    model?: string,
  ): Promise<EvaluationResponse | null> {
    const actionPayload = "actionType" in action
      ? {
          action_type: (action as ActionPayload).actionType,
          target_system: (action as ActionPayload).targetSystem,
          payload_summary: (action as ActionPayload).payloadSummary,
          delegation_depth: (action as ActionPayload).delegationDepth ?? 0,
        }
      : action;
    return this.request<EvaluationResponse>("POST", "/api/v1/engine/evaluate", {
      action: actionPayload,
      intent_id: intentId,
      trace_id: traceId,
      model,
    });
  }

  // ── Observability ──────────────────────────────────────────────

  async ingestSpan(
    traceId: string,
    spanType: string,
    data: Record<string, unknown>,
    agentId = "",
    intentId = "",
  ): Promise<SpanResponse | null> {
    return this.request<SpanResponse>("POST", "/api/v1/observability/traces", {
      trace_id: traceId,
      span_type: spanType,
      data,
      agent_id: agentId,
      intent_id: intentId,
    });
  }

  async analyzeTrace(traceId: string): Promise<AnalyzeTraceResponse | null> {
    return this.request<AnalyzeTraceResponse>("POST", `/api/v1/observability/traces/${traceId}/analyze`);
  }

  // ── Audit ──────────────────────────────────────────────────────

  async verifyArtifact(artifactId: string): Promise<ArtifactVerifyResponse | null> {
    return this.request<ArtifactVerifyResponse>("GET", `/api/v1/audit/verify/${artifactId}`);
  }

  async exportCompliance(): Promise<ComplianceExportResponse | null> {
    return this.request<ComplianceExportResponse>("GET", "/api/v1/audit/export");
  }

  // ── Health ─────────────────────────────────────────────────────

  async health(): Promise<boolean> {
    try {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.timeout);
      const res = await fetch(`${this.baseUrl}/health`, {
        headers: this.headers,
        signal: controller.signal,
      });
      clearTimeout(timer);
      return res.status === 200;
    } catch {
      return false;
    }
  }

  // ── Telemetry ──────────────────────────────────────────────────

  async getTelemetryStats(): Promise<TelemetryStatsResponse | null> {
    return this.request<TelemetryStatsResponse>("GET", "/api/v1/telemetry/stats");
  }

  async getAgentTelemetry(
    agentId: string,
    limit = 100,
  ): Promise<TelemetryRecord[] | null> {
    return this.request<TelemetryRecord[]>("GET", `/api/v1/telemetry/agent/${agentId}?limit=${limit}`);
  }

  async getAgentStats(agentId: string): Promise<AgentStatsResponse | null> {
    return this.request<AgentStatsResponse>("GET", `/api/v1/telemetry/agent/${agentId}/stats`);
  }

  async getTraceTelemetry(traceId: string): Promise<TelemetryRecord[] | null> {
    return this.request<TelemetryRecord[]>("GET", `/api/v1/telemetry/trace/${traceId}`);
  }

  // ── Prompt Analyzer ────────────────────────────────────────────

  async analyzePrompt(
    prompt: string,
    intentId?: string,
    agentId?: string,
  ): Promise<PromptAnalysisResponse | null> {
    return this.request<PromptAnalysisResponse>(
      "POST",
      "/api/v1/prompt/analyze",
      { prompt, intent_id: intentId, agent_id: agentId },
      120_000,
    );
  }

  async analyzePromptBatch(
    prompts: string[],
    intentId?: string,
    agentId?: string,
  ): Promise<PromptBatchResponse | null> {
    return this.request<PromptBatchResponse>(
      "POST",
      "/api/v1/prompt/analyze/batch",
      { prompts, intent_id: intentId, agent_id: agentId },
      300_000,
    );
  }

  // ── JIT Credentials ────────────────────────────────────────────

  async mintCredential(params: MintCredentialParams): Promise<CredentialResponse | null> {
    return this.request<CredentialResponse>("POST", "/api/v1/auth/credentials/mint", {
      intent_id: params.intentId,
      artifact_id: params.artifactId,
      scope: params.scope ?? {},
      ...(params.ttlSeconds ? { ttl_seconds: params.ttlSeconds } : {}),
    });
  }

  async revokeCredential(
    credentialId: string,
    reason = "task_complete",
  ): Promise<Record<string, unknown> | null> {
    return this.request("POST", "/api/v1/auth/credentials/revoke", {
      credential_id: credentialId,
      reason,
    });
  }

  async validateCredential(token: string): Promise<CredentialResponse | null> {
    return this.request<CredentialResponse>("POST", "/api/v1/auth/credentials/validate", { token });
  }

  // ── PBAC Policies ──────────────────────────────────────────────

  async evaluatePbac(
    intentId: string,
    actionType: string,
    targetSystem = "",
  ): Promise<PbacEvalResponse | null> {
    return this.request<PbacEvalResponse>("POST", "/api/v1/auth/policies/evaluate", {
      intent_id: intentId,
      action_type: actionType,
      target_system: targetSystem,
    });
  }

  async listPolicies(intentId?: string): Promise<PbacPolicyResponse[] | null> {
    const qs = intentId ? `?intent_id=${intentId}` : "";
    return this.request<PbacPolicyResponse[]>("GET", `/api/v1/auth/policies${qs}`);
  }

  // ── Agent DID Registry ────────────────────────────────────────

  async registerAgentDID(params: {
    agentId: string;
    displayName?: string;
  }): Promise<DIDResponse | null> {
    return this.request<DIDResponse>("POST", "/api/v1/dids/register", {
      agent_id: params.agentId,
      display_name: params.displayName,
    });
  }

  async resolveDID(did: string): Promise<DIDResponse | null> {
    return this.request<DIDResponse>("GET", `/api/v1/dids/${did}`);
  }

  async revokeDID(
    did: string,
    reason = "cessation",
  ): Promise<DIDResponse | null> {
    return this.request<DIDResponse>("POST", `/api/v1/dids/${did}/revoke`, { reason });
  }

  async rotateDIDKey(did: string): Promise<DIDKeyRotateResponse | null> {
    return this.request<DIDKeyRotateResponse>("POST", `/api/v1/dids/${did}/keys/rotate`);
  }
}

File: python/scouter/api/backend.py
"""
Scouter Backend HTTP Client.

Provides typed methods to communicate with the Scouter backend API.
Used by ScouterClient when ``backend_url`` is configured.
Falls back gracefully — returns None on connection errors so the
SDK can fall back to local evaluation.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional
from urllib.parse import quote

import httpx

logger = logging.getLogger("scouter.backend")


def _enc(value: str) -> str:
    """Encode a value safely for inclusion in a URL path segment."""
    return quote(str(value), safe="")


class CapabilityEscalationError(Exception):
    """Raised when a credential mint is blocked by aggregate scope ceiling or dangerous capability combination."""

    def __init__(self, detail: dict):
        self.detail = detail
        super().__init__(detail.get("reason", "Capability escalation"))


class BackendClient:
    """HTTP client for the Scouter backend REST API."""

    def __init__(
        self,
        base_url: str,
        timeout: float = 5.0,
        api_key: Optional[str] = None,
        client: Optional[Any] = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        headers = {}
        if api_key:
            headers["X-Scouter-API-Key"] = api_key
        self._client = client or httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            headers=headers,
            limits=httpx.Limits(
                max_keepalive_connections=10,
                max_connections=20,
                keepalive_expiry=30,
            ),
        )

    def _request(self, method: str, path: str, *, caller: str = "", **kwargs) -> Optional[dict]:
        """Unified request handler with proper error logging."""
        try:
            r = self._client.request(method, path, **kwargs)
            r.raise_for_status()
            return r.json()
        except httpx.TimeoutException:
            logger.warning("Scouter backend timeout on %s %s (%s)", method, path, caller)
            return None
        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            if status == 409:
                # Capability escalation — surface to caller instead of swallowing
                try:
                    detail = e.response.json()
                except Exception:
                    detail = {"reason": e.response.text}
                raise CapabilityEscalationError(detail)
            if status == 401 or status == 403:
                logger.error("Scouter backend auth failed (%d) on %s %s — check API key", status, method, path)
            elif status >= 500:
                logger.warning("Scouter backend error (%d) on %s %s (%s)", status, method, path, caller)
            else:
                logger.debug("Scouter backend %d on %s %s (%s)", status, method, path, caller)
            return None
        except httpx.HTTPError as e:
            logger.warning("Scouter backend connection error on %s %s: %s", method, path, e)
            return None

    # ── Intent Registry ────────────────────────────────────────────

    def register_intent(
        self,
        agent_id: str,
        natural_language: str,
        permitted_actions: List[str],
        excluded_actions: List[str],
        principal_chain: Optional[List[dict]] = None,
        version: str = "1.0",
    ) -> Optional[dict]:
        return self._request("POST", "/api/v1/intents", caller="register_intent", json={
            "agent_id": agent_id,
            "natural_language": natural_language,
            "permitted_actions": permitted_actions,
            "excluded_actions": excluded_actions,
            "principal_chain": principal_chain or [],
            "version": version,
        })

    def get_intent(self, intent_id: str) -> Optional[dict]:
        return self._request("GET", f"/api/v1/intents/{_enc(intent_id)}", caller="get_intent")

    # ── Consequence Engine ─────────────────────────────────────────

    def evaluate(
        self,
        action: Dict[str, Any],
        intent_id: Optional[str] = None,
        trace_id: Optional[str] = None,
        model: Optional[str] = None,
    ) -> Optional[dict]:
        return self._request("POST", "/api/v1/engine/evaluate", caller="evaluate", json={
            "action": action,
            "intent_id": intent_id,
            "trace_id": trace_id,
            "model": model,
        })

    # ── Observability ──────────────────────────────────────────────

    def ingest_span(
        self,
        trace_id: str,
        span_type: str,
        data: dict,
        agent_id: str = "",
        intent_id: str = "",
    ) -> Optional[dict]:
        return self._request("POST", "/api/v1/observability/traces", caller="ingest_span", json={
            "trace_id": trace_id,
            "span_type": span_type,
            "data": data,
            "agent_id": agent_id,
            "intent_id": intent_id,
        })

    def analyze_trace(self, trace_id: str) -> Optional[dict]:
        return self._request("POST", f"/api/v1/observability/traces/{_enc(trace_id)}/analyze", caller="analyze_trace")

    # ── Audit ──────────────────────────────────────────────────────

    def verify_artifact(self, artifact_id: str) -> Optional[dict]:
        return self._request("GET", f"/api/v1/audit/verify/{_enc(artifact_id)}", caller="verify_artifact")

    def export_compliance(self) -> Optional[dict]:
        return self._request("GET", "/api/v1/audit/export", caller="export_compliance")

    # ── Health ─────────────────────────────────────────────────────

    def health(self) -> bool:
        try:
            r = self._client.get("/health")
            return r.status_code == 200
        except httpx.HTTPError:
            return False

    # ── Telemetry ──────────────────────────────────────────────────

    def get_telemetry_stats(self) -> Optional[dict]:
        """Get system-wide telemetry statistics."""
        return self._request("GET", "/api/v1/telemetry/stats", caller="get_telemetry_stats")

    def get_agent_telemetry(
        self, agent_id: str, limit: int = 100
    ) -> Optional[List[dict]]:
        """Get telemetry records for a specific agent."""
        return self._request("GET", f"/api/v1/telemetry/agent/{_enc(agent_id)}", caller="get_agent_telemetry", params={"limit": limit})

    def get_agent_stats(self, agent_id: str) -> Optional[dict]:
        """Get aggregated statistics for an agent."""
        return self._request("GET", f"/api/v1/telemetry/agent/{_enc(agent_id)}/stats", caller="get_agent_stats")

    def get_trace_telemetry(self, trace_id: str) -> Optional[List[dict]]:
        """Get telemetry records for a specific trace."""
        return self._request("GET", f"/api/v1/telemetry/trace/{_enc(trace_id)}", caller="get_trace_telemetry")

    # ── Prompt Analyzer ────────────────────────────────────────────

    def analyze_prompt(
        self,
        prompt: str,
        intent_id: Optional[str] = None,
        agent_id: Optional[str] = None,
    ) -> Optional[dict]:
        """Analyze a prompt for intent, risk, impact, and get an ALLOWED/REJECTED decision."""
        return self._request("POST", "/api/v1/prompt/analyze", caller="analyze_prompt", json={
            "prompt": prompt,
            "intent_id": intent_id,
            "agent_id": agent_id,
        }, timeout=120.0)

    def analyze_prompt_batch(
        self,
        prompts: List[str],
        intent_id: Optional[str] = None,
        agent_id: Optional[str] = None,
    ) -> Optional[dict]:
        """Analyze multiple prompts in a single request."""
        return self._request("POST", "/api/v1/prompt/analyze/batch", caller="analyze_prompt_batch", json={
            "prompts": prompts,
            "intent_id": intent_id,
            "agent_id": agent_id,
        }, timeout=300.0)

    # ── JIT Credentials ────────────────────────────────────────────

    def mint_credential(
        self,
        intent_id: str,
        artifact_id: str,
        scope: Optional[dict] = None,
        ttl_seconds: Optional[int] = None,
        task_id: Optional[str] = None,
    ) -> Optional[dict]:
        """Mint a JIT credential after a PASS_THROUGH decision."""
        body: Dict[str, Any] = {
            "intent_id": intent_id,
            "artifact_id": artifact_id,
            "scope": scope or {},
        }
        if ttl_seconds:
            body["ttl_seconds"] = ttl_seconds
        if task_id:
            body["task_id"] = task_id
        return self._request("POST", "/api/v1/auth/credentials/mint", caller="mint_credential", json=body)

    # ── Task Lifecycle (Change 2) ──────────────────────────────────

    def open_task(
        self,
        intent_id: Optional[str] = None,
        agent_id: Optional[str] = None,
        description: Optional[str] = None,
        parent_task_id: Optional[str] = None,
    ) -> Optional[dict]:
        """Open a task. Bind credentials minted within it to its task_id."""
        return self._request(
            "POST", "/api/v1/auth/tasks/open", caller="open_task",
            json={
                "intent_id": intent_id,
                "agent_id": agent_id,
                "description": description,
                "parent_task_id": parent_task_id,
            },
        )

    def close_task(self, task_id: str, terminal_event: str = "completed") -> Optional[dict]:
        """Close a task — cascades revocation to every bound credential."""
        return self._request(
            "POST", "/api/v1/auth/tasks/close", caller="close_task",
            json={"task_id": task_id, "terminal_event": terminal_event},
        )

    def checkpoint_task(self, task_id: str) -> Optional[dict]:
        """Heartbeat for long-running tasks. Resets staleness timer."""
        return self._request(
            "POST", "/api/v1/auth/tasks/checkpoint", caller="checkpoint_task",
            json={"task_id": task_id},
        )

    def get_task(self, task_id: str) -> Optional[dict]:
        return self._request("GET", f"/api/v1/auth/tasks/{_enc(task_id)}", caller="get_task")

    def revoke_credential(
        self,
        credential_id: str,
        reason: str = "task_complete",
    ) -> Optional[dict]:
        """Revoke a JIT credential."""
        return self._request("POST", "/api/v1/auth/credentials/revoke", caller="revoke_credential", json={
            "credential_id": credential_id, "reason": reason,
        })

    def validate_credential(self, token: str) -> Optional[dict]:
        """Validate a JIT credential token."""
        return self._request("POST", "/api/v1/auth/credentials/validate", caller="validate_credential", json={
            "token": token,
        })

    # ── PBAC Policies ──────────────────────────────────────────────

    def evaluate_pbac(
        self,
        intent_id: str,
        action_type: str,
        target_system: str = "",
    ) -> Optional[dict]:
        """Check PBAC policy for an action."""
        return self._request("POST", "/api/v1/auth/policies/evaluate", caller="evaluate_pbac", json={
            "intent_id": intent_id,
            "action_type": action_type,
            "target_system": target_system,
        })

    def list_policies(self, intent_id: Optional[str] = None) -> Optional[List[dict]]:
        """List PBAC policies."""
        params = {}
        if intent_id:
            params["intent_id"] = intent_id
        return self._request("GET", "/api/v1/auth/policies", caller="list_policies", params=params)

    # ── Agent DID Registry ─────────────────────────────────────────

    def register_agent_did(
        self, agent_id: str, display_name: Optional[str] = None
    ) -> Optional[dict]:
        """Register a DID for an agent. Returns DID Document + private key."""
        return self._request("POST", "/api/v1/dids/register", caller="register_agent_did", json={
            "agent_id": agent_id,
            "display_name": display_name,
        })

    def resolve_did(self, did: str) -> Optional[dict]:
        """Resolve a DID Document by DID string."""
        return self._request("GET", f"/api/v1/dids/{_enc(did)}", caller="resolve_did")

    def revoke_did(self, did: str, reason: str = "cessation") -> Optional[dict]:
        """Revoke a DID. Invalidates associated JIT credentials."""
        return self._request("POST", f"/api/v1/dids/{_enc(did)}/revoke", caller="revoke_did", json={
            "reason": reason,
        })

    def rotate_did_key(self, did: str) -> Optional[dict]:
        """Rotate the Ed25519 keypair for a DID."""
        return self._request("POST", f"/api/v1/dids/{_enc(did)}/keys/rotate", caller="rotate_did_key")
